src/linear_algebra/matrix/diag.lean

```diff
@@ -0,0 +1,226 @@
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d) :=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h]}
+end
+
+/-- Every unit matrix is diagonal. -/
+@[simp] lemma is_diag_unit [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+by { intros i j h, have h' := subsingleton.elim i j, contradiction }
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, by simp
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+by { intros i j h, simp [h] }
+
+@[simp] lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0) :
+  (A.map f).is_diag :=
+by { intros i j h, simp [ha h, hf] }
+
+@[simp] lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by {intros i j h, simp [ha h]}
+
+@[simp] lemma is_diag.add
+  [add_zero_class α] {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A + B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.sub [add_group α]
+  {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A - B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
+  {k : R} {A : matrix n n α} (ha : A.is_diag) :
+  (k • A).is_diag :=
+by { intros i j h, simp [ha h] }
+
+@[simp] lemma is_diag.transpose
+  [has_zero α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᵀ.is_diag :=
+by { intros i j h, simp [ha h.symm] }
+
+@[simp] lemma is_diag.conj_transpose
+  [semiring α] [star_ring α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᴴ.is_diag :=
+by { intros i j h, simp [ha h.symm] }
+
+@[simp] lemma is_diag.minor [has_zero α]
+  {A : matrix n n α} (ha : A.is_diag) {f : m → n} (hf : injective f) :
+  (A.minor f f).is_diag :=
+by { intros i j h, specialize @hf i j, simp [ha (mt hf h)] }
+
+/-- `(A ⊗ B).is_diag` if both `A` and `B` are diagonal. -/
+@[simp] lemma is_diag.kronecker [mul_zero_class α]
+  {A : matrix m m α} {B : matrix n n α} (hA: A.is_diag) (hB: B.is_diag) :
+  (A ⊗ₖ B).is_diag :=
+begin
+  rintros ⟨a, b⟩ ⟨c, d⟩ h,
+  dsimp [kronecker_apply],
+  by_cases hac: a = c,
+  { have hbd: b ≠ d, {tidy}, simp [hB hbd] },
+  { simp [hA hac] },
+end
+
+@[simp] lemma is_diag.diagonal [has_zero α] [decidable_eq n] (d : n → α) :
+  (diagonal d).is_diag :=
+by { intros i j h, simp [h] }
+
+lemma is_diag.is_symm [has_zero α] {A : matrix n n α} (h : A.is_diag) :
+  A.is_symm :=
+begin
+  ext,
```

Comment:
```suggestion
  ext i j,
```
The rest of your proof mentions this auto-named variables. The proof becomes more stable if you specify the name explicitly. (If in the future the tactic `ext` would generate a different name, then this proof would break.)